### PR TITLE
support quic streams per priority

### PR DESCRIPTION
**Pull request by wyfo**
Links' internal design has been modified to take priority in arguments. For link supporting priorities, like Quic, read/write have been split by priority.
Quic implement stream per priority using a single bidirectional stream for control messages (and opening), and unidirectional streaming for the other priorities.

---
## 🏷️ Label-Based Checklist

Based on the labels applied to this PR, please complete these additional requirements:

**Labels:** `new feature`

## 🆕 New Feature Requirements

Since this PR adds a new feature:

- [x] **Feature scope documented** - Clear description of what the feature does and why it's needed
- [x] **Minimum necessary code** - Implementation is as simple as possible, doesn't overcomplicate the system
- [x] **New APIs well-designed** - Public APIs are intuitive, consistent with existing APIs
- [x] **Comprehensive tests** - All functionality is tested (happy path + edge cases + error cases)
- [x] **Examples provided** - Usage examples in code comments or separate example files
- [x] **Documentation added** - New docs explaining the feature, its use cases, and API
- [x] **Feature flag considered** - Can the feature be enabled/disabled for gradual rollout?
- [x] **Performance impact assessed** - Memory, CPU, storage implications measured
- [x] **Integration tested** - Feature works with existing features

**Consider:** Can this feature be split into smaller, incremental PRs?

**Instructions:**
1. Check off items as you complete them (change `- [ ]` to `- [x]`)
2. The PR checklist CI will verify these are completed

*This checklist updates automatically when labels change, but preserves your checked boxes.*

### CODE REVIEW

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 1]
```diff
@@ -13,68 +13,159 @@
...
+                PROTOCOL_SINGLE_STREAM.into(),
+                PROTOCOL_LEGACY.into(),
+            ],
+        }
+    }
+
+    fn max_concurrent_uni_streams(&self) -> quinn::VarInt {
+        match self {
+            Self::Disabled => 0u8.into(),
+            _ => (Priority::NUM as u8 - 1).into(),
```
Comment (oteffahi) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs:96]: Please comment to explain that a bidirectional stream exists for priority 0

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 2]
```diff
@@ -13,68 +13,224 @@
...
+    /// `multistream=true`
+    Enabled,
+    /// default, or `multistream=auto`
+    Auto,
+}
+
+impl MultiStreamConfig {
+    /// Parse multistream configuration.
+    fn new(config: Config) -> ZResult<Self> {
+        Ok(match config.get("multistream") {
```
Comment (oteffahi) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs]: Maybe this key should be exposed in `zenoh_link_commons` so other links that will support multistream will use the same parameter.

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 3]
```diff
@@ -13,68 +13,224 @@
...
+        mut priority_txs: HashMap<usize, oneshot::Sender<quinn::RecvStream>>,
+    ) -> ZResult<()> {
+        while !priority_txs.is_empty() {
+            let recv = connection.accept_uni().await?;
+            // Uni streams' indexes starts from zero, while priorities above Control starts from 1,
+            // hence the `+ 1`
+            let prio = recv.id().index() as usize + 1;
+            if let Some(tx) = priority_txs.remove(&prio) {
+                tx.send(recv).ok();
+            }
```
Comment (oteffahi) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs:173]: I think there's a potential silent failure here. Maybe we should handle the `else` and log an error.

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 4]
```diff
@@ -83,13 +239,56 @@ impl LinkUnicastQuic {
...
+    /// Retrieved the write-stream mapped to the priority
+    ///
+    /// # Safety
+    ///
+    /// There should be only one caller per priority.
+    #[allow(clippy::mut_from_ref)]
+    unsafe fn write_stream(&self, priority: Priority) -> &mut quinn::SendStream {
+        unsafe { &mut *self.send[priority as usize].get() }
+            .as_mut()
+            .expect("multistream should have been started")
```
Comment (oteffahi) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs:262]: This error message would be confusing if it occurs in single-stream mode.

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 5]
```diff
@@ -83,13 +239,56 @@ impl LinkUnicastQuic {
...
+    /// The stream may be pending, in which case we wait until it is accepted.
+    ///
+    /// # Safety
+    ///
+    /// There should be only one caller per priority.
+    #[allow(clippy::mut_from_ref)]
+    async unsafe fn read_stream(&self, priority: Priority) -> ZResult<&mut quinn::RecvStream> {
+        match unsafe { &mut *self.recv[priority as usize].get() }
+            .as_mut()
+            .expect("multistream should have been started")
```
Comment (oteffahi) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs:279]: This error message would be confusing if it occurs in single-stream mode.

File: io/zenoh-transport/src/multicast/link.rs [hunk 6]
```diff
@@ -37,14 +37,13 @@ use crate::{
...
-            TransmissionPipeline, TransmissionPipelineConf, TransmissionPipelineConsumer,
-            TransmissionPipelineProducer,
+            PipelineConsumer, TransmissionPipeline, TransmissionPipelineConf,
+            TransmissionPipelineConsumer, TransmissionPipelineProducer,
         },
         priority::TransportPriorityTx,
     },
     multicast::transport::TransportMulticastInner,
 };
-
```
Comment (oteffahi) [inline io/zenoh-transport/src/multicast/link.rs:45]: Please add this line-break back so I can sleep at night :')

File: io/zenoh-transport/src/unicast/universal/link.rs [hunk 7]
```diff
@@ -184,59 +197,108 @@ impl TransportLinkUnicastUniversal {
...
-                            stats.inc_tx_bytes(n);
-                        }
-                    }
+            pull = pipeline.pull() => {
+                let Some((mut batch, priority)) = pull else {
+                    // The queue has been disabled: break the tx loop, drain the queue, and exit
+                    break
+                };
+                debug_assert!(write_priority == Priority::Control || priority == write_priority);
+                link.send_batch(&mut batch, write_priority).await?;
```
Comment (oteffahi) [inline io/zenoh-transport/src/unicast/universal/link.rs]: I think it's better here to `link.send_batch(&mut batch, priority)` and let the underlying link decide the scheduling based on if it is set to multistream or not. Writing everything under `Priority::Control` in single-stream links is a bit confusing, and could be cause for false assumptions later if links use the message's priority for features other than stream multiplexing.

Will the additional check on every write have a non-negligible impact on performance? Or I guess we can have some additional abstraction to avoid this check.

File: io/zenoh-transport/src/unicast/universal/link.rs [hunk 8]
```diff
@@ -256,50 +318,167 @@ async fn rx_task(
...
+    if link.link.supports_priorities() {
+        let (res, _, _) = select_all((Priority::MAX as u8..Priority::MIN as u8).map(|prio| {
+            let mut link = link.clone();
+            let transport = transport.clone();
+            let token = token.clone();
+            let lease_tracker = lease_tracker.clone();
+            #[cfg(feature = "stats")]
+            let stats = stats.clone();
+            let pool = pool.clone();
+            zenoh_runtime::ZRuntime::RX.spawn(async move {
```
Comment (oteffahi) [inline io/zenoh-transport/src/unicast/universal/link.rs:377]: I think we can use `tokio::runtime::Handle::current()` to match the behavior of the `else` which schedules the future on whatever runtime called the `rx_task`.
I believe the same remark applies for `tx_task` and its `write_loop` futures.

File: io/zenoh-transport/src/unicast/universal/link.rs [hunk 9]
```diff
@@ -256,50 +318,167 @@ async fn rx_task(
...
+            }
+            Poll::Pending
+        })
+        .await
+    }
+}
+
+impl Drop for TimeoutTracker {
+    fn drop(&mut self) {
+        if Arc::strong_count(&self.0) == 2 {
```
Comment (oteffahi) [inline io/zenoh-transport/src/unicast/universal/link.rs]: Couldn't this be avoided by passing a `Weak` to the task? Or can the upgrade call take too long and impact the lease?

File: io/zenoh-transport/src/unicast/link.rs [hunk 10]
```diff
@@ -92,7 +92,7 @@ impl TransportLinkUnicast {
 
     pub(crate) async fn send(&self, msg: &TransportMessage) -> ZResult<usize> {
         let mut link = self.tx();
-        link.send(msg).await
+        link.send(msg, Priority::Control).await
```
Comment (oteffahi) [inline io/zenoh-transport/src/unicast/link.rs]: May be worth to comment that this is only used by the Zenoh handshake or to close the link, and therefore can be hardcoded to `Priority::Control`.

File: io/zenoh-transport/src/unicast/universal/link.rs (hunk 7 above)
Comment (wyfo) [reply io/zenoh-transport/src/unicast/universal/link.rs]: I made the priority optional to avoid confusion

File: io/zenoh-transport/src/unicast/universal/link.rs (hunk 8 above)
Comment (wyfo) [reply io/zenoh-transport/src/unicast/universal/link.rs:377]: Everywhere else, RX runtime is used

File: io/zenoh-transport/src/unicast/universal/link.rs (hunk 9 above)
Comment (wyfo) [reply io/zenoh-transport/src/unicast/universal/link.rs]: done

File: io/zenoh-transport/src/unicast/link.rs (hunk 10 above)
Comment (wyfo) [reply io/zenoh-transport/src/unicast/link.rs]: Priority is optional now

File: io/zenoh-transport/src/multicast/link.rs (hunk 6 above)
Comment (wyfo) [reply io/zenoh-transport/src/multicast/link.rs:45]: done

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs (hunk 5 above)
Comment (wyfo) [reply io/zenoh-links/zenoh-link-quic/src/unicast.rs:279]: It should not occurs in single-stream mode

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs (hunk 4 above)
Comment (wyfo) [reply io/zenoh-links/zenoh-link-quic/src/unicast.rs:262]: It should not occurs in single-stream mode

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs (hunk 3 above)
Comment (wyfo) [reply io/zenoh-links/zenoh-link-quic/src/unicast.rs:173]: I added a comment

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs (hunk 2 above)
Comment (wyfo) [reply io/zenoh-links/zenoh-link-quic/src/unicast.rs]: Done

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs (hunk 3 above)
Comment (oteffahi) [reply io/zenoh-links/zenoh-link-quic/src/unicast.rs:173]: My problem isn't about the channel, it's about the `if let Some(tx) = priority_txs.remove(&prio)`. If the `remove` is expected to always return `Some`, then it should be clear in the code and panic/error-out if this guarantee is lost in the future. An `if let Some` without an `else` here feels to me like a ticking bomb.

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 11]
```diff
@@ -13,68 +13,226 @@
...
+impl MultiStreamConfig {
+    /// Parse multistream configuration.
+    fn new(config: Config) -> ZResult<Self> {
+        Ok(match config.get(MULTISTREAM_CONFIG_KEY) {
+            Some("false") => Self::Disabled,
+            Some("true") => Self::Enabled,
+            None | Some("auto") => Self::Auto,
+            Some(s) => bail!("Invalid multistream config: {s}"),
+        })
+    }
```
Comment (oteffahi) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs]: This should be an endpoint parameter instead of an endpoint config. (`quic/127.0.0.1:7447?multistream=0` instead of `quic/127.0.0.1:7447#multistream=0`). It would enable the listener's multistream mode to be propagated in scouting.

**Comment by oteffahi** [comment]
@wyfo I have pushed tests for this PR. Please review them.

**Comment by wyfo** [comment]
LGTM

**Review by oteffahi — APPROVED** [review]

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 12]
```diff
@@ -83,13 +245,59 @@ impl LinkUnicastQuic {
     async fn close(&self) -> ZResult<()> {
         tracing::trace!("Closing QUIC link: {}", self);
         // Flush the QUIC stream
-        let mut guard = zasynclock!(self.send);
-        if let Err(e) = guard.finish() {
-            tracing::trace!("Error closing QUIC stream {}: {}", self, e);
+        for (i, send) in self.send.iter().enumerate() {
+            if let Some(send) = unsafe { &mut (*send.get()) }.as_mut() {
```
Comment (oteffahi) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs:243]: Correct me if I'm wrong, but since `LinkUnicastTrait::close` can be called concurrently with `LinkUnicastTrait::write` and `LinkUnicastTrait::write_all`, the `UnsafeCell` usage here is a UB.

Maybe we can skip closing the streams and just close the connection?

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 13]
```diff
@@ -325,6 +544,8 @@ impl LinkManagerUnicastTrait for LinkManagerUnicastQuic {
             })?
             .await
             .map_err(|e| zerror!("Can not create a new QUIC link bound to {}: {}", host, e))?;
+        quic_conn.set_max_concurrent_bi_streams(1u8.into());
+        quic_conn.set_max_concurrent_uni_streams(multistream.max_concurrent_uni_streams());
```
Comment (oteffahi) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs]: This changes the config after the QUIC handshake, by sending dedicated messages. These limits should be set on the `quinn::TransportConfig` passed to `quic_endpoint.set_default_client_config`.
Line 529 can be changed like so:
```rust
        quic_endpoint.set_default_client_config({
            let mut client_config = quinn::ClientConfig::new(Arc::new(quic_config));
            let mut transport_config = quinn::TransportConfig::default();
            
            transport_config.max_concurrent_bidi_streams(1u8.into());
            transport_config.max_concurrent_uni_streams(multistream.max_concurrent_uni_streams());
            
            client_config.transport_config(transport_config.into());
            client_config
        });
```

**Review by oteffahi — CHANGES_REQUESTED** [review]

File: io/zenoh-link-commons/src/quic.rs [hunk 14]
```diff
@@ -42,8 +42,13 @@ use crate::{
     ConfigurationInspector, LinkAuthId, BIND_INTERFACE,
 };
 
-// Default ALPN protocol
-pub const ALPN_QUIC_HTTP: &[&[u8]] = &[b"hq-29"];
+// ALPN protocols
+/// Protocol used by zenoh <= 1.6.2
```
Comment (oteffahi) [inline io/zenoh-link-commons/src/quic.rs]:
```suggestion
/// Protocol used by zenoh <= 1.8.0
```

File: io/zenoh-links/zenoh-link-quic/src/unicast.rs [hunk 15]
```diff
@@ -13,68 +13,230 @@
...
+impl MultiStreamConfig {
+    /// Parse multistream configuration.
+    fn new(metadata: Metadata) -> ZResult<Self> {
+        let multistream = metadata.get(Metadata::MULTISTREAM).unwrap_or("auto");
+        if multistream == "auto" {
+            return Ok(Self::Auto);
+        }
+        if multistream
+            .parse()
+            .map_err(|_| zerror!("Invalid multistream config:  {multistream}"))?
```
Comment (oteffahi) [inline io/zenoh-links/zenoh-link-quic/src/unicast.rs]: My bad, I though boolean parsing would work for 0/1 values, but it turns out it doesn't 😅. Please change this back to checking for "0" and "1".

**Review by oteffahi — CHANGES_REQUESTED** [review]
